### PR TITLE
storage2: add pebble-based payload storage

### PR DESCRIPTION
**Pull request by SaveTheRbtz**
This PR adds Pebble-based MVCC storage for Payloads.

It layouts payloads on disk using the following suffix encoding:
```
{Owner, "/", Key, "/", EncodedHeight}
```
Where `EncodedHeight` is the height's 1's compliment encoded in Big Endian.  This way we can use [SeekPrefixGE](https://pkg.go.dev/github.com/cockroachdb/pebble#Iterator.SeekPrefixGE) API (which, as name implies, does not support reverse iteration).
`SeekPrefixGE` has an advantage over the `SeekGE` that it uses bloom filter and reduce read IO.

Values are encoded "as is" and we rely on database block-level compression and caching.

>  Note to reviewers: PR looks quite big, but it is mostly MVCC tests and benchmarks, for the core code there are only ~100 LOCs in [service/storage2/payload/payload.go](https://github.com/onflow/flow-archive/pull/104/files#diff-2032581afe217d4fbf3abd6b2a09524cb9f0db23b33bebf786da0e791f577ca3).

ref: https://github.com/onflow/flow-go/pull/3937
ref: https://github.com/onflow/flow-archive/issues/99
ref: "[CockroachDB MVCC](https://github.com/cockroachdb/pebble/blob/master/docs/RFCS/20211018_range_keys.md)".

depends on: https://github.com/onflow/flow-archive/pull/105

### CODE REVIEW

File: service/storage2/payload/payload.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+) ([]byte, error) {
+	iter := s.db.NewIter(&pebble.IterOptions{
+		UseL6Filters: true,
+	})
+	defer iter.Close()
+
+	encoded := newLookupKey(height, reg).Bytes()
+	ok := iter.SeekPrefixGE(encoded)
+	if !ok {
+		return nil, &ErrNotFound{
```
Comment (pattyshack) [inline service/storage2/payload/payload.go:63]: in general, there's no such thing as not found in flow.  not found implies empty string

File: service/storage2/payload/payload.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+// NewPayloadStorage creates a pebble-backed payload storage.
+// The reason we use a separate storage for payloads we need a Comparer with a custom Split function.
+//
+// It needs to access the last available payload with height less or equal to the requested height.
+// This means all point-lookups are range scans.
+func NewStorage(dbPath string, cacheSize int64) (*payloadStorage, error) {
+	// A custom Split function splits the key into the prefix and the suffix.
+	// This is needed for SeekPrefixGE to work.
+	comparer := *pebble.DefaultComparer
```
Comment (pattyshack) [inline service/storage2/payload/payload.go:29]: maybe refactor out the comparer and add test.

**Review by pattyshack — COMMENTED** [review]
lgtm in general, except for how not found is being handled

File: service/storage2/payload/lookup.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
+package payload
+
+import (
+	"encoding/binary"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+type lookupKey struct {
```
Comment (zhangchiqing) [inline service/storage2/payload/lookup.go:9]:
```suggestion
// lookupKey is the encoded format of the storage key for looking up register value
type lookupKey struct {
```

File: service/storage2/config.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+		L0StopWritesThreshold:       1000,
+		LBaseMaxBytes:               64 << 20, // 64 MB
+		Levels:                      make([]pebble.LevelOptions, 7),
+		MaxOpenFiles:                16384,
+		MemTableSize:                64 << 20,
+		MemTableStopWritesThreshold: 4,
+		MaxConcurrentCompactions:    func() int { return 3 },
+	}
+
+	for i := 0; i < len(opts.Levels); i++ {
```
Comment (zhangchiqing) [inline service/storage2/config.go:23]: is this for the boomfilter? can you explain how these configs work?

File: service/storage2/payload/payload.go (hunk 1 above)
Comment (SaveTheRbtz) [reply service/storage2/payload/payload.go:64]: Yeah, this makes sense in merkle trie, not sure whether we need to carry the same assumption to the database-baked register storage.  Let me sleep on it.

Comment (SaveTheRbtz) [reply service/storage2/payload/payload.go]: Yeah, you are right.  By doing this we'll be able to simplify this codepath: https://github.com/onflow/flow-archive/blob/6f8333901d5539e7d5c6f84e86d20f7f80dfbf27/service/index/reader.go#L120-L125

Comment (pattyshack) [reply service/storage2/payload/payload.go]: This actually has nothing to do with merkle tree in general.  This is a flow specific design choice that we are stuck with (having not found error in the merkle tree would have actually make things easier).  Returning not found error is actually a bug.

**Review by pattyshack — APPROVED** [review]

File: service/storage2/config.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+		l := &opts.Levels[i]
+		// The default is 4KiB (uncompressed), which is too small
+		// for good performance (esp. on stripped storage).
+		l.BlockSize = 32 << 10       // 32 KB
+		l.IndexBlockSize = 256 << 10 // 256 KB
+
+		// 10 bits per key yields a filter with <1% false positive rate.
```
Comment (zhangchiqing) [inline service/storage2/config.go:40]:
```suggestion
		// 10 bits per key yields a filter with <1% false positive rate.
		// the bloom filter is required to support SeekPrefixGE used for getting 
		// most recent value of a ledger register.
```

File: service/storage2/payload/lookup.go [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+// lookupKey is the encoded format of the storage key for looking up register value
+type lookupKey struct {
+	encoded []byte
+}
+
+func newLookupKey(height uint64, reg flow.RegisterID) *lookupKey {
+	key := lookupKey{
+		encoded: make([]byte, 0, len(reg.Owner)+1+len(reg.Key)+1+heightSuffixLen),
+	}
+
```
Comment (zhangchiqing) [inline service/storage2/payload/lookup.go:18]:
```suggestion

	// the lookup key is mostly used in a way to find most recent value for a register.
	// the "<owner>/<key>" part is the register key, which is used as a prefix to filter and iterate 
	// through updated values at different heights, and find the most recent updated value up to 
	// a certain height.
```

**Review by zhangchiqing — APPROVED** [review]

**Pull request merged**